File: rust/tests/magnetometer/as5600_test/src/main.rs
use embedded_hal::i2c::I2c as _;
use linux_embedded_hal::I2cdev;
use periph::chips::magnetometer::As5600Full;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x36);

    let mut dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");

    println!("--- magnet status (60 s max) ---");
    let deadline = std::time::Instant::now() + std::time::Duration::from_secs(60);
    loop {
        let mut buf = [0u8; 1];
        dev.write_read(addr, &[0x0Bu8], &mut buf).ok();
        let s = buf[0];
        dev.write_read(addr, &[0x1Au8], &mut buf).ok();
        let agc = buf[0];
        let md = s & 0x08 != 0; let ml = s & 0x10 != 0; let mh = s & 0x20 != 0;
        println!("MD={} ML={} MH={} AGC={}", md as u8, ml as u8, mh as u8, agc);
        if md || std::time::Instant::now() >= deadline { break; }
        std::thread::sleep(std::time::Duration::from_millis(200));
    }
    println!("--- end magnet status ---");

    let mut chip = As5600Full::new(dev, addr).expect("init AS5600");

    let mut passed = 0i32;
    let mut failed = 0i32;

    // --- Magnet detection ---
    check_true!(chip.is_magnet_detected().unwrap_or(false), "magnet_detected", passed, failed);

    // --- Angle readings ---
    let a = chip.angle().unwrap_or(0.0);
    check_true!(a >= 0.0 && a < 360.0, "angle_range", passed, failed);

    let r = chip.angle_raw().unwrap_or(0);
    check_true!(r <= 4095, "angle_raw_range", passed, failed);

    let ra = chip.raw_angle().unwrap_or(0);
    check_true!(ra <= 4095, "raw_angle_range", passed, failed);

    let rad = chip.raw_angle_degrees().unwrap_or(0.0);
    check_true!(rad >= 0.0 && rad < 360.0, "raw_angle_degrees_range", passed, failed);

    // --- Diagnostics ---
    check_true!(chip.agc().unwrap_or(0) >= 0, "agc_non_negative", passed, failed);
    check_true!(chip.magnitude().unwrap_or(0) >= 0, "magnitude_non_negative", passed, failed);

    // --- Status ---
    let sb = chip.status_byte().unwrap_or(0);
    check_true!(sb <= 255, "status_byte_valid", passed, failed);

    // --- Position configuration (volatile) ---
    chip.set_zero_position(100).unwrap();
    check_true!(chip.zero_position().unwrap() == 100, "zero_position_after_set", passed, failed);

    chip.set_max_position(2000).unwrap();
    check_true!(chip.max_position().unwrap() == 2000, "max_position_after_set", passed, failed);

    chip.set_max_angle(2048).unwrap();
    check_true!(chip.max_angle().unwrap() == 2048, "max_angle_after_set", passed, failed);

    // --- Configure ---
    chip.configure(0, 0, 0, 0, 0, 0, false).unwrap();
    check_true!(chip.is_magnet_detected().unwrap_or(false), "configure_accepted", passed, failed);

    // --- Burn count ---
    let bc = chip.burn_count().unwrap_or(0);
    check_true!(bc <= 3, "burn_count_range", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/magnetometer/as5600_test/main.go
//go:build linux && !tinygo

// AS5600 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the AS5600 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/magnetometer"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x36"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := magnetometer.NewAs5600Full(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	// --- Magnet detection ---
	md, err := chip.IsMagnetDetected()
	check("magnet_detected", err == nil && md)

	// --- Angle readings ---
	a, err := chip.Angle()
	check("angle_range", err == nil && a >= 0.0 && a < 360.0)
	r, err := chip.AngleRaw()
	check("angle_raw_range", err == nil && r <= 4095)
	ra, err := chip.RawAngle()
	check("raw_angle_range", err == nil && ra <= 4095)
	rad, err := chip.RawAngleDegrees()
	check("raw_angle_degrees_range", err == nil && rad >= 0.0 && rad < 360.0)

	// --- Diagnostics ---
	_, err = chip.AGC()
	check("agc_ok", err == nil)
	mag, err := chip.Magnitude()
	check("magnitude_non_negative", err == nil && mag >= 0)

	// --- Status ---
	sb, err := chip.StatusByte()
	check("status_byte_valid", err == nil && sb <= 255)

	// --- Position configuration (volatile) ---
	if err := chip.SetZeroPosition(100); err != nil {
		fmt.Fprintln(os.Stderr, "set_zero:", err)
	}
	zp, _ := chip.ZeroPosition()
	check("zero_position_after_set", zp == 100)

	if err := chip.SetMaxPosition(2000); err != nil {
		fmt.Fprintln(os.Stderr, "set_max:", err)
	}
	mp, _ := chip.MaxPosition()
	check("max_position_after_set", mp == 2000)

	if err := chip.SetMaxAngle(2048); err != nil {
		fmt.Fprintln(os.Stderr, "set_max_angle:", err)
	}
	ma, _ := chip.MaxAngle()
	check("max_angle_after_set", ma == 2048)

	// --- Configure ---
	if err := chip.Configure(0, 0, 0, 0, 0, 0, false); err != nil {
		fmt.Fprintln(os.Stderr, "configure:", err)
	}
	md2, _ := chip.IsMagnetDetected()
	check("configure_accepted", md2)

	// --- Burn count ---
	bc, err := chip.BurnCount()
	check("burn_count_range", err == nil && bc <= 3)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
